demos/spectre_v2.cc

```diff
@@ -0,0 +1,298 @@
+  }
+};
+
+// Fetches an address N-times in a real reader (FirstActualRead or
+// SecondActualRead).
+template<unsigned int N> struct Fetches {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(const void *addr) {
+    (void)*reinterpret_cast<const volatile char *>(addr);
+    Fetches<N - 1>::generate(addr);
+  }
+};
+
+template<> struct Fetches<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(const void *addr) {
+    // Empty.
+  }
+};
+
+// Repeatedly indirectly invokes a real reader on a safe address in
+// LeakByte.
+template<unsigned int N> struct SafeCalls {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (*real_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL *real_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *safe_address) {
+    real_reader(safe_address);
+    SafeCalls<N - 1>::generate(real_reader, safe_address);
+  }
+};
+
+template<> struct SafeCalls<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (*real_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL *real_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *safe_address) {
+    // Empty.
+  }
+};
+
+// Repeatedly flushes a dummy reader from the cache and invokes it on an
+// unsafe address in LeakByte.
+template<unsigned int N> struct DummyCalls {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (**dummy_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL **dummy_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *unsafe_address) {
+    CLFlush(dummy_reader);
+    (*dummy_reader)(unsafe_address);
+    DummyCalls<N - 1>::generate(dummy_reader, unsafe_address);
+  }
+};
+
+template<> struct DummyCalls<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (**dummy_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL **dummy_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *unsafe_address) {
+    // Empty.
+  }
+};
+
+// We fastcall the readers in the 32-bit mode so that there is only one
+// instruction. Otherwise there would be two (fetch the pointer from the stack
+// and fetch from the pointer).
+void
+#if SAFESIDE_IA32
+SAFESIDE_FASTCALL
+#endif
+FirstActualRead(const void *addr) {
+  // We read the value 100 times to have enough places to jump to.
+  Fetches<100>::generate(addr);
+}
+
+// DummyRead is between the two real reads, so that misjump chances are the
+// highest possible.
+void
+#if SAFESIDE_IA32
+SAFESIDE_FASTCALL
+#endif
+DummyRead(const void *addr) {
+  // Does nothing - only NOPs.
+  Nops<100>::generate();
```

Comment:
Reading this file top-to-bottom, I have no idea why this function is made of 100 NOPs. I see the explanation comes later, in `LeakByte`.

For the benefit of future readers, I think we should have an explanation of what we're doing that includes justification for the NOPs _before_ we define this function -- either a comment immediately attached, or as further detail in the toplevel file comment.